lib/mfg.c

```diff
@@ -136,6 +136,49 @@ static float spi_clk_hi_rate_float[] = {
 	120, 80, 60, 48, 40, 34, 30, 26.67, 24, 21.82
 };
 
+int switchtec_security_spi_avail_rate_get(struct switchtec_dev *dev,
+		struct switchtec_security_spi_avail_rate *rates)
+{
+	int ret;
+	struct cfg_reply {
+		uint32_t valid;
+		uint32_t rsvd1;
+		uint64_t cfg;
+		uint32_t public_key_exponent;
+		uint8_t rsvd2;
+		uint8_t public_key_num;
+		uint8_t public_key_ver;
+		uint8_t spi_core_clk_high;
+		uint8_t public_key[SWITCHTEC_KMSK_NUM][SWITCHTEC_KMSK_LEN];
+		uint8_t rsvd4[32];
+	} reply;
+	uint8_t subcmd = 0;
+
+	if (switchtec_gen(dev) == SWITCHTEC_GEN5) {
+		subcmd = 1;
+		ret = switchtec_mfg_cmd(dev,
+					MRPC_SECURITY_CONFIG_GET_GEN5,
+					&subcmd, sizeof(subcmd),
+					&reply, sizeof(reply));
+	} else {
+		ret = switchtec_mfg_cmd(dev, MRPC_SECURITY_CONFIG_GET,
+					NULL, 0, &reply,
+					sizeof(reply));
+	}
+	if (ret)
+		return ret;
```

Comment:
Looks like a bunch of this should be factored out of secure_config_get() into a shared common function instead of copied and pasted. Should help to cleanup secure_config_get() as well given how complicated that function has gotten.

Reply:
Thanks for the suggestion. Code is now much cleaner with the common function.